receiver/valkeyreceiver/client.go

```diff
@@ -0,0 +1,67 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+package valkeyreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/valkeyreceiver"
+
+import (
+	"context"
+	"strings"
+
+	"github.com/valkey-io/valkey-go"
+)
+
+var attrDelimiter = "\r\n"
+
+// Interface for a Valkey client. Implementation can be faked for testing.
+type client interface {
+	// retrieves a string of key/value pairs of valkey metadata
+	retrieveInfo(context.Context) (map[string]string, error)
+	// close release Valkey client connection pool
+	close() error
+}
+
+// Wraps a real Valkey client, implements `client` interface.
+type valkeyClient struct {
+	client valkey.Client
+}
+
+var _ client = (*valkeyClient)(nil)
+
+// Creates a new real Valkey client from the passed-in valkey.Options.
+func newValkeyClient(options valkey.ClientOption) (client, error) {
+	innerClient, err := valkey.NewClient(options)
+	if err != nil {
+		return nil, err
+	}
+	return &valkeyClient{innerClient}, nil
+}
+
+// Retrieve Valkey INFO. We retrieve all of the 'sections'.
+func (c *valkeyClient) retrieveInfo(ctx context.Context) (map[string]string, error) {
+	str, err := c.client.Do(ctx, c.client.B().Info().Build()).ToString()
+	if err != nil {
+		return nil, err
+	}
+
+	return parseRawDataMap(str), nil
+}
+
+func parseRawDataMap(data string) map[string]string {
+	attrs := make(map[string]string)
+	lines := strings.Split(data, attrDelimiter)
+	for _, line := range lines {
+		if len(line) == 0 || strings.HasPrefix(line, "#") {
+			continue
+		}
+		pair := strings.Split(line, ":")
```

Comment:
Should use SplitN, as values (example from unit test `listener0:name=tcp,bind=*,bind=-::*,port=6379`) can have `:` characters.

Reply:
Not sure about this one, it looks like the keys available in the info output do not contain `:` characters. For the shared example, `listener0` is the key, while `name=tcp,bind=*,bind=-::*,port=6379` is value. This parsing function is only interested in capturing the keys and any value string (can contain `:`).

Reply:
Changed to `Cut` to only split on the first delimiter instance: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37005/commits/1a76c237301b3515c50dbeab0b27a9ddb4a00883